index/generator/library/util.go

```diff
@@ -0,0 +1,356 @@
+package library
+
+import (
+	"archive/zip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/devfile/library/pkg/testingutil/filesystem"
+	dfutil "github.com/devfile/library/pkg/util"
+	"github.com/devfile/registry-support/index/generator/schema"
+	gitpkg "github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+)
+
+// DownloadRemoteStack downloads the stack version outside of the registry repo
+func DownloadRemoteStack(git *schema.Git, path string, verbose bool) (err error) {
+
+	// convert revision to referenceName type, ref name could be a branch or tag
+	// if revision is not specified it would be the default branch of the project
+	revision := git.Revision
+	refName := plumbing.ReferenceName(git.Revision)
+
+	if plumbing.IsHash(revision) {
+		// Specifying commit in the reference name is not supported by the go-git library
+		// while doing git.PlainClone()
+		fmt.Printf("Specifying commit in 'revision' is not yet supported.")
+		// overriding revision to empty as we do not support this
+		revision = ""
+	}
+
+	if revision != "" {
+		// lets consider revision to be a branch name first
+		refName = plumbing.NewBranchReferenceName(revision)
+	}
+
+	cloneOptions := &gitpkg.CloneOptions{
+		URL:           git.Url,
+		RemoteName:    git.RemoteName,
+		ReferenceName: refName,
+		SingleBranch:  true,
+		// we don't need history for starter projects
+		Depth: 1,
+	}
+
+	originalPath := ""
+	if git.SubDir != "" {
+		originalPath = path
+		path, err = ioutil.TempDir("", "")
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err = gitpkg.PlainClone(path, false, cloneOptions)
+
+	if err != nil {
+
+		// it returns the following error if no matching ref found
+		// if we get this error, we are trying again considering revision as tag, only if revision is specified.
+		if _, ok := err.(gitpkg.NoMatchingRefSpecError); !ok || revision == "" {
+			return err
+		}
+
+		// try again to consider revision as tag name
+		cloneOptions.ReferenceName = plumbing.NewTagReferenceName(revision)
+		// remove if any .git folder downloaded in above try
+		_ = os.RemoveAll(filepath.Join(path, ".git"))
+		_, err = gitpkg.PlainClone(path, false, cloneOptions)
+		if err != nil {
+			return err
+		}
+	}
+
+	// we don't want to download project be a git repo
+	err = os.RemoveAll(filepath.Join(path, ".git"))
+	if err != nil {
+		// we don't need to return (fail) if this happens
+		fmt.Printf("Unable to delete .git from cloned devfile repository")
+	}
+
+	if git.SubDir != "" {
+		err = GitSubDir(path, originalPath,
+			git.SubDir)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+
+}
+
+// DownloadStackFromGit downloads the stack from a git repo
+func DownloadStackFromGit(git *schema.Git, path string, verbose bool) ([]byte, error) {
```

Comment:
Would be good to add a comment to the function header explaining how this function is different from `DownloadRemoteStack`